File: CSharp/LC101-Unit2/Class-2.1/Lecture.cs
using System;

/**
 * LC 101 - Unit 2 
 * Class 2.1
 * Chapter 2 - Data Types
 * Hank DeDona
 * 9/29/20
 */
namespace Class_2._1
{
    class Program
    {
        static void Main(string[] args)
        {
			// Single line comment
			/*
			 * Block comment
			 */

			Console.WriteLine("2.1.1. Static vs. Dynamic Typing");
			/*
			 * C# is a statically typed language. When a variable or parameter is declared
			 * in a statically typed language, the data type for the value must be
			 * specified. Once the declaration is made, the variable or parameter cannot
			 * refer to a value of any other type
			 */
			string myDogsName = "Milo";

			// Nope!
			// myDogsName = 42; // This is trying to store an int in a String, legal in
			// javascript, not in C#

			Console.WriteLine("2.1.2 Built-In Types");

			// There are a bunch of built-in types: int, float, double, char and bool - These are primitives (and immutable)
			// There are also .NET class equivalents: Int32, Single, Double, Char, Boolean, String - These are classes

			Console.WriteLine("2.1.2.3. Operations");

			// Note: Operators like + and * are type-dependant. For example when + is used with Strings they concatenate,
			// when + is used with numbers they add the numbers

			Console.WriteLine("Hello" + " class!"); // Concatenate
			Console.WriteLine(5 + 3);   // Addition

			// Note: You can mix and match number types, the output will be the more precise (in this case double)
			float a = 2;
			double b = 3;
			Console.WriteLine((a + b).GetType());   // Double

			Console.WriteLine("2.1.3. Reference and Value Types");
			// Note: ALL primitives are Value types

			Console.WriteLine("2.1.3.1. Class Types");

			Cat myCat = new Cat();  // Declare the variable myCat of type Cat, the "= new" instantiates an instance

			Cat myNamedCat = new Cat("Garfield");   // Constructor that takes in a name

			int catAge = 11;	// Get's it's own memory with the value 11
			Cat sameCat = myCat;        // What's happening here? sameCat references the same memory as myCat

			Console.WriteLine(myCat.Equals(sameCat));       // What do you think the output will be?

			Console.WriteLine("2.1.3.2. Boxing/Unboxing");

			int primitiveIntVal = 5;    // Primitive
			Console.WriteLine("Primitive Int Val: " + primitiveIntVal);

			Int32 objectIntVal = primitiveIntVal;   // Boxing (into a class object provided a primitive)
			Console.WriteLine("Primitive Int Object: " + objectIntVal);

			int samePrimitiveIntVal = objectIntVal; // Unboxing (into a primitive provided a class object)
			Console.WriteLine("Primitive Int Val (copy): " + samePrimitiveIntVal);

			// Note: Console is part of the System package, since we said "using System" above we don't have to do
			//System.Console.WriteLine("This is unnecessary because of our using above");

			// The following code below is from the csharp-web-dev-lsn1datatypes solution
			string input;
			Double fahrenheit;

			// 2.2.4. Input / Output and the Console Class
			Console.WriteLine("Temperature in F:");     // Output
			input = Console.ReadLine();                 // Input a String
			fahrenheit = double.Parse(input);           // Convert to a double

			Console.WriteLine("You input: " + fahrenheit);

			// 2.3. Strings, Characters, and Arrays
			Console.WriteLine("2.3.1.2. Single vs. Double Quotation Marks");

			string stringVar = "Dog"; // Strings require double quotation marks
			char charVar = 'd';       // Characters use single quotation marks

			Console.WriteLine("2.3.1.3. String Manipulation");

            string str = "Rutabaga";

			Console.WriteLine(str.Substring(3, 1));     // Returns the character in 3rd position, (a).
			Console.WriteLine(str.Substring(2, 3));     // Return substring from 2nd to 4th, i.e. substring starting at index 2 and 3 characters long, (tab).
			Console.WriteLine(str.Length);              // Tells us the length of the string, (8).
			Console.WriteLine(str.IndexOf('a'));        // Returns the index for the first occurrence of ‘a’, (3).
			Console.WriteLine(str.Split('a')[0] + ", " + str.Split('a')[1] + ", " + str.Split('a')[2]);  // Splits the string into sections at each delimiter and stores the sections as elements in an array, ({Rut, b, g}).
			Console.WriteLine(str + str);               // Concatenate two strings together, (RutabagaRutabaga).
			Console.WriteLine(str.Trim());              // Removes any whitespace at the beginning or end of the string, (Rutabaga — there’s not whitespace here).
			Console.WriteLine(str.ToUpper());           // Changes all alphabetic characters in the string to UPPERCASE
			Console.WriteLine(str.ToLower());           // Changes all alphabetic characters in the string to lowercase


			Console.WriteLine("2.3.2. Arrays");

			int[] someInts = new int[10];       // Creates an array of size 10 that contains integers 

			int[] someOtherInts = { 1, 1, 2, 3, 5, 8 };     // Creates an array of integers with 1,1,2,3,5,8 stored

			int anInt = someOtherInts[4];
			Console.WriteLine(anInt);       // Should output 5, remember arrays are 0 based

			// As we’ve said several times now, every variable in C# refers to an object!!

			Console.WriteLine("2.4.2 Static Methods");
			// Refer to Cat.makeNoise() method

			// Notice that we didn't have to instantiate it first! 
			Cat.makeNoise(null);        // Outputs roar!
			Cat.makeNoise("purr");      // Outputs purr

			// 2.4.3. HelloMethods
			// Refer to HelloMethods
		}
	}
}

File: Java/Class-2.1/src/org/launchcode/main/Lecture.java
package org.launchcode.main;

import java.util.Scanner;

public class Lecture {

	public static void main(String[] args) throws CloneNotSupportedException {

		// Single line comment
		/*
		 * Block comment
		 */

		System.out.println("2.1.1. Static vs. Dynamic Typing");
		/*
		 * Java is a statically typed language. When a variable or parameter is declared
		 * in a statically typed language, the data type for the value must be
		 * specified. Once the declaration is made, the variable or parameter cannot
		 * refer to a value of any other type
		 */
		String myDogsName = "Milo";

		// Nope!
		// myDogsName = 42; // This is trying to store an int in a String, legal in
		// javascript, not in java

		System.out.println("2.1.2.2. Single vs. Double Quotation Marks");
		String staticVariable = "dog"; // String
		char charVariable = 'd'; // Single Char
		// Nope!
		// charVariable = 'ab';

		System.out.println("2.1.2.3. String Manipulation");
		String str = "Rutabaga";
		System.out.println(str.charAt(3));
		System.out.println(str.substring(2, 4));
		System.out.println(str.length());
		int startIndex = (str.indexOf('a')+1);
		System.out.println(str.indexOf('a', startIndex));

		System.out.println("2.1.3. Primitive Types");
		int answerToAllThings = 42;
		float pi = 3.141593f; // Note: need to add an f at the end of floats!
		double piDouble = 3.14159265358979;
		boolean iAmTrue = true;

		System.out.println("2.1.4. Non-primitive Types");
		String cool = "LaunchCode";

		System.out.println("2.1.5. Autoboxing");

		// Autoboxing primitive -> object
		int somePrimitiveInteger = 5;
		Integer someIntegerObject = somePrimitiveInteger;

		// Autounboxing object -> primitive
		int someNewPrimitiveInt = someIntegerObject;

		System.out.println("2.2.6. Collect Input with the Scanner Class");
		Scanner input = new Scanner(System.in); // System.in is from the console (user input)
		System.out.print("Enter the temperature in Fahrenheit: ");
		double fahrenheit = input.nextDouble();	// Wait for user input
		double celsius = (fahrenheit - 32) * 5 / 9;

		System.out.println("The temperature in Celsius is: " + celsius);

		input.close();	// Close the scanner!!

		System.out.println("2.3.1. Arrays");
		// Array of 10 ints (empty array)
		int[] someInts = new int[10];

		// Initialize an array with 6 values on a single line (1,1,2,3,5,8)
		int[] someOtherInts = { 1, 1, 2, 3, 5, 8 };

		// To access the data in the array someOtherInts, we use square brackets
		int someVal = someOtherInts[4]; // This will store 5 in someVal

		System.out.println("2.3.2. Java Objects");
		System.out.println("Repeat after me, everything that is not a primitive is an object in Java!");
		System.out.println("Everything that is not a primitive is an object in Java!");

		String language = "Java";

		// The String class is an object, we can call methods on objects.
		System.out.println(language.length()); // This will print the length of the string language (in this case 4)

		System.out.println("2.3.3. Class Types");
		/*
		 * A class is a template for creating objects. In addition to the object types
		 * introduced so far, any class in Java also defines a type. We’ll have much
		 * more to say about classes and objects, but for now you need to recognize the
		 * basic syntax of class types and class creation.
		 */
		
		// Create an instance of the Cat class called myDefaultCat using the NEW keyword
		// 1) Cat myDefaultCat declares the variable myCat and sets it to be of type Cat.
		// 2) = new Cat() initializes the variable with a new Cat object.
		// 3) Any arguments that are required to build the new Cat object must be included within the parentheses.
		Cat myDefaultCat = new Cat();
		
		Cat myNamedCat = new Cat("KitKat");	// Including values within the parenthesis
		
		System.out.println("2.3.3.1. Reference Variables");
		
		int firstCatAge = 11;
		int secondCatAge = firstCatAge;	// firstCatAge and secondCatAge are separate values (11)
		
		Cat myCat = new Cat("buster");
		Cat sameCat = myCat;	// myCat and sameCat point to the same values in memory
		
		System.out.println("MyCat name: " + myCat.getName());
		System.out.println("SameCat name: " + sameCat.getName());
		
		myCat.setName("blah");
		
		System.out.println("MyCat name: " + myCat.getName());
		System.out.println("SameCat name: " + sameCat.getName());
		
		// Lets make sameCat the same VALUE (pass by value, NOT reference!)
		
		Cat cloneCat = myCat.clone();
		
		cloneCat.setName("Clonie");
		
		System.out.println("MyCat name: " + myCat.getName());
		System.out.println("Clonecat name: " + cloneCat.getName());
		
		System.out.println("2.3.4. Static Methods");
		// Static methods can be called on a class directly without instatiating a variable first
		Cat.makeNoise();	// Cat is the class, not an object (like myCat or sameCat)
	}

}
